File: androidAlekt/src/main/java/com/app/alektapp/android/purchases/Subscription.kt
package com.app.alektapp.android.purchases

import android.content.Context
import com.app.alektapp.android.R
import com.app.alektapp.android.analitic.AppFlyer
import com.app.alektapp.domain.model.subscriptionModel.SubscriptionProduct
import com.app.alektapp.domain.model.subscriptionModel.SubscriptionStatus
import com.app.alektapp.domain.services.SubscriptionService
import com.revenuecat.purchases.CacheFetchPolicy
import com.revenuecat.purchases.CustomerInfo
import com.revenuecat.purchases.LogLevel
import com.revenuecat.purchases.Package
import com.revenuecat.purchases.PurchaseParams
import com.revenuecat.purchases.Purchases
import com.revenuecat.purchases.PurchasesConfiguration
import com.revenuecat.purchases.PurchasesError
import com.revenuecat.purchases.getOfferingsWith
import com.revenuecat.purchases.interfaces.LogInCallback
import com.revenuecat.purchases.interfaces.ReceiveCustomerInfoCallback
import com.revenuecat.purchases.purchaseWith
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine

class Subscription(val context: Context):
    SubscriptionService {

    private val appFluer = AppFlyer(context.getString(R.string.AppFlyer))

    private val key = context.getString(R.string.RevenueKey)

    private var products: List<Package> = listOf()

    override fun setDebug(isDebug: Boolean) {
        Purchases.logLevel =  if (isDebug) LogLevel.DEBUG else LogLevel.ERROR

        appFluer.setDebug(isDebug)
    }

    override fun init() {
        Purchases.configure(PurchasesConfiguration.Builder(context, key).build())

        appFluer.initService(context)
    }

    override fun init(userId: String) {
        Purchases.configure(PurchasesConfiguration.Builder(context, key).appUserID(userId).build())

        appFluer.initService(context)
    }

    override suspend fun logIn(userId: String): Boolean = suspendCoroutine { con ->
        Purchases.sharedInstance.logIn(userId, object : LogInCallback {
            override fun onError(error: PurchasesError) = con.resume(false)

            override fun onReceived(customerInfo: CustomerInfo, created: Boolean) = con.resume(true)

        })

        appFluer.logIn(userId)
    }

    override suspend fun isSubscribed(): SubscriptionStatus = suspendCoroutine { con ->
        Purchases.sharedInstance.getCustomerInfo(CacheFetchPolicy.default(), object : ReceiveCustomerInfoCallback {
            override fun onError(error: PurchasesError) = con.resume(SubscriptionStatus.Err(error.message))

            override fun onReceived(customerInfo: CustomerInfo) {
                with (customerInfo.entitlements.active) {
                    when {
                        isNotEmpty() -> con.resume(SubscriptionStatus.IsSubscribed)
                        isEmpty() -> con.resume(SubscriptionStatus.NotSubscribed)
                        else -> con.resume(SubscriptionStatus.Err(""))
                    }
                }
            }
        })
    }

    override suspend fun fetchOffers(): Set<SubscriptionProduct> = suspendCoroutine { cor ->
        Purchases.sharedInstance.getOfferingsWith({ error ->
            cor.resume(setOf())
        }) { offerings ->
            offerings.current?.availablePackages?.let {
                    offers ->
                products = offers
                offers.map { it.toBase() }.run {
                    cor.resume(this.toSet())
                }
            }
        }
    }

    override suspend fun subscribe(context: Any, productBase: SubscriptionProduct): SubscriptionStatus = suspendCoroutine { cor ->
        products.find { it.product.title == productBase.title }?.let {
            (context as Context).findActivity()?.let { activity ->
                Purchases.sharedInstance.purchaseWith(
                    PurchaseParams.Builder(activity, it).build(),
                onSuccess = { _, _ ->
                    cor.resume(SubscriptionStatus.IsSubscribed)

                    appFluer.trackPurchase(context, productBase)
                    },
                    onError = { err, _ ->
                        cor.resume(SubscriptionStatus.Err(err.message))
                    }
                )
            }
        }
    }
}

File: iosAlekt/iosAlekt/Subscription/Subscription.swift

//
//  Subscription.swift
//  iosAlectApp
//
//  Created by 111 on 06.05.2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import RevenueCat
import shared

class Subscription: SubscriptionService {
    
    private let appFlyer: AppFlyer
    
    private var key: String {
        if let path = Bundle.main.path(forResource: "Config", ofType: "plist") {
            if let dict = NSDictionary(contentsOfFile: path) as? [String: AnyObject] {
                if let stringValue = dict["RevenueKay"] as? String {
                    return stringValue
                }
            }
        }
        fatalError("no key for RevenueCat")
    }

    private let entitlementId = "basic_motch"

    private var offers = [Package]()
    
    init(appFlyer: AppFlyer) {
        self.appFlyer = appFlyer
        self.offers = []
    }

    func setDebug(isDebug: Bool) {
        guard isDebug else { return }
        Purchases.logLevel = .debug
    }

    func doInit() {
        Purchases.configure(withAPIKey: key)
    }

    func doInit(userId: String) {
        Purchases.configure(withAPIKey: key, appUserID: userId)
    }

    func isSubscribed() async throws -> SubscriptionStatus {
        do {
            let info = try await Purchases.shared.customerInfo()
            
                if !info.entitlements.active.isEmpty {
                    return SubscriptionStatus.IsSubscribed()
                } else {
                    return SubscriptionStatus.NotSubscribed()
                }
        } catch {
            return SubscriptionStatus.Err(err: error.localizedDescription)
        }
    }
    
    func fetchOffers() async throws -> Set<SubscriptionProduct> {
        
        do {
            let offers = try await Purchases.shared.offerings()
                        
            if let packages = offers.current?.availablePackages {
                self.offers = packages
                
                return Set(packages.map {  $0.toBase() })
            }
            
            print("offerings is empty")
            return Set()
            
        } catch {
            print(error.localizedDescription)
            return Set()
        }
    }
   

    func logIn(userId: String) async throws -> KotlinBoolean {
        do {
            let success = try await Purchases.shared.logIn(userId)
            
            return KotlinBoolean(bool: success.created)
        } catch {
            print(error)
            return false
        }
    }

    func subscribe(context: Any, productBase: SubscriptionProduct) async throws -> SubscriptionStatus {
            let package = offers.first { $0.storeProduct.localizedTitle == productBase.title}
        
            guard let package else {
                return SubscriptionStatus.Err(err: "package not fpond")
            }
            
        do {
            let result = try await Purchases.shared.purchase(package: package)
            
            if result.customerInfo.entitlements[self.entitlementId]?.isActive == true {
                appFlyer.trackPurcahse(offer: productBase)
                return SubscriptionStatus.IsSubscribed()
          }
            
            return SubscriptionStatus.Err(err: "purcahses can not subscribe")
        } catch {
            return SubscriptionStatus.Err(err: error.localizedDescription)
        }
    }
}

extension Package {
    func toBase() -> SubscriptionProduct {
        var type = PackageTypeBase.unknown

        switch self.packageType {
        case .unknown: type = PackageTypeBase.unknown
        case .custom: type = PackageTypeBase.custom
        case .lifetime: type = PackageTypeBase.lifetime
        case .annual: type = PackageTypeBase.annual
        case .sixMonth: type = PackageTypeBase.sixMonth
        case .threeMonth: type = PackageTypeBase.twoMonth
        case .twoMonth: type = PackageTypeBase.twoMonth
        case .monthly: type = PackageTypeBase.monthly
        case .weekly: type = PackageTypeBase.weekly
        }

        return SubscriptionProduct(
            price: self.storeProduct.localizedPriceString,
            title: self.storeProduct.localizedTitle,
            description: self.storeProduct.localizedDescription,
            type: type
       )
    }
}
